exp/addons/controllers/clusterresourceset_controller.go

```diff
@@ -0,0 +1,357 @@
+
+// ClusterResourceSetReconciler reconciles a ClusterResourceSet object
+type ClusterResourceSetReconciler struct {
+	Client client.Client
+	Log    logr.Logger
+
+	scheme *runtime.Scheme
+}
+
+func (r *ClusterResourceSetReconciler) SetupWithManager(mgr ctrl.Manager, options controller.Options) error {
+	c, err := ctrl.NewControllerManagedBy(mgr).
+		For(&addonsv1.ClusterResourceSet{}).
+		Watches(
+			&source.Kind{Type: &clusterv1.Cluster{}},
+			&handler.EnqueueRequestsFromMapFunc{ToRequests: handler.ToRequestsFunc(r.clusterToClusterResourceSet)},
+		).
+		WithOptions(options).
+		WithEventFilter(predicates.ResourceNotPaused(r.Log)).
+		Build(r)
+	if err != nil {
+		return errors.Wrap(err, "failed setting up with a controller manager")
+	}
+
+	err = c.Watch(
+		&source.Kind{Type: &clusterv1.Cluster{}},
+		&handler.EnqueueRequestsFromMapFunc{
+			ToRequests: handler.ToRequestsFunc(r.clusterToClusterResourceSet),
+		},
+		// TODO: should this wait for Cluster.Status.InfrastructureReady similar to Infra Machine resources?
+		predicates.ClusterUnpaused(r.Log),
+	)
+	if err != nil {
+		return errors.Wrap(err, "failed adding Watch for Cluster to ClusterResourceSet controller manager")
+	}
+
+	r.scheme = mgr.GetScheme()
+	return nil
+}
+
+func (r *ClusterResourceSetReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, reterr error) {
+	ctx := context.Background()
+
+	// Fetch the ClusterResourceSet instance.
+	clusterResourceSet := &addonsv1.ClusterResourceSet{}
+	if err := r.Client.Get(ctx, req.NamespacedName, clusterResourceSet); err != nil {
+		if apierrors.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	// Initialize the patch helper.
+	patchHelper, err := patch.NewHelper(clusterResourceSet, r.Client, patch.WithStatusObservedGeneration{})
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	defer func() {
+		// Always attempt to Patch the ClusterResourceSet object and status after each reconciliation.
+		if err := patchHelper.Patch(ctx, clusterResourceSet); err != nil {
+			reterr = kerrors.NewAggregate([]error{reterr, err})
+		}
+	}()
+
+	// Handle normal reconciliation loop.
+	return r.reconcile(ctx, clusterResourceSet)
+}
+
+// reconcile handles cluster reconciliation.
+func (r *ClusterResourceSetReconciler) reconcile(ctx context.Context, clusterResourceSet *addonsv1.ClusterResourceSet) (ctrl.Result, error) {
+	logger := r.Log.WithValues("ClusterResourceSet", clusterResourceSet.Name, "namespace", clusterResourceSet.Namespace)
+
+	clusters, err := r.getClustersByClusterResourceSetSelector(ctx, clusterResourceSet)
+	if err != nil {
+		logger.Error(err, "Failed fetching clusters that matches ClusterResourceSet labels", "ClusterResourceSet", clusterResourceSet.Name)
```

Comment:
We probably need to set a condition here. Anything we log is not readily and easily visible to the end user.